libraries/lib-basic-ui/BasicUI.h

```diff
@@ -0,0 +1,289 @@
+{
+public:
+   virtual ~ProgressDialog();
+
+   // Update the bar and poll for clicks
+   virtual ProgressResult Poll(
+      unsigned long long numerator,
+      unsigned long long denominator,
+      const TranslatableString &message = {}) = 0;
+};
+
+//! Abstraction of a progress dialog with undefined time-to-completion estimate
+class BASIC_UI_API GenericProgressDialog
+{
+public:
+   virtual ~GenericProgressDialog();
+   //! Give some visual indication of progress
+   virtual void Pulse() = 0;
+};
+
+//! @}
+
+//! Abstract class defines a few user interface services, not mentioning particular toolkits
+/*! The intention is that the application supplies a concrete implementation at
+ startup.  Most code will not use this class directly, but call the inline
+ functions that follow. */
+class BASIC_UI_API Services {
+public:
+   virtual ~Services();
+   virtual void DoCallAfter(const Action &action) = 0;
+   virtual void DoYield() = 0;
+   virtual void DoShowErrorDialog(const WindowPlacement &placement,
+      const TranslatableString &dlogTitle,
+      const TranslatableString &message,
+      const ManualPageID &helpPage,
+      const ErrorDialogOptions &options) = 0;
+   virtual MessageBoxResult DoMessageBox(
+      const TranslatableString& message,
+      MessageBoxOptions options) = 0;
+   virtual std::unique_ptr<ProgressDialog>
+   DoMakeProgress(const TranslatableString &title,
+      const TranslatableString &message,
+      unsigned flag,
+      const TranslatableString &remainingLabelText) = 0;
+   virtual std::unique_ptr<GenericProgressDialog>
+   DoMakeGenericProgress(const WindowPlacement &placement,
+      const TranslatableString &title,
+      const TranslatableString &message) = 0;
+};
+
+//! Fetch the global instance, or nullptr if none is yet installed
+BASIC_UI_API Services *Get();
+
+//! Install an implementation; return the previously installed instance
+BASIC_UI_API Services *Install(Services *pInstance);
+
+//! @name Functions that dispatch to the global Services, if supplied, or else are no-ops
+//! @{
+
+//! Schedule an action to be done later in the event loop, and in the main thread
+inline void CallAfter(const Action &action)
+{
+   if (auto p = Get())
+      p->DoCallAfter(action);
+}
+
+//! Let the main event loop dispatch waiting events, including those enqueued by CallAfter
+inline void Yield()
+{
+   if (auto p = Get())
+      p->DoYield();
+}
+
+//! Show an error dialog with a link to the manual for further help
+inline void ShowErrorDialog(
+   const WindowPlacement &placement,    //!< how to parent the dialog
+   const TranslatableString &dlogTitle, //!< Text for title bar
+   const TranslatableString &message,   //!< The main message text
+   const ManualPageID &helpPage,          //!< Identifies manual page (and maybe an anchor)
+   const ErrorDialogOptions &options = {})
+{
+   if (auto p = Get())
+      p->DoShowErrorDialog(placement, dlogTitle, message, helpPage, options);
```

Comment:
Same here, client code may attempt to show an error while UI doesn't exist.

Reply:
Unlike with the case of Action, this is a void-valued function that doesn't promise any side effect.  I think it is all right to have a default implementation that is no-op.